grpc_unary_attestation/src/client.rs

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::{unary_session_client::UnarySessionClient, UnaryRequest};
+use anyhow::Context;
+use oak_functions_abi;
+use oak_remote_attestation::handshaker::{
+    AttestationBehavior, ClientHandshaker, Encryptor, ServerIdentityVerifier,
+};
+use tonic::transport::Channel;
+
+type SessionId = u64;
+
+/// gRPC Attestation Service client implementation.
+pub struct AttestationClient {
+    session_id: SessionId,
+    encryptor: Encryptor,
+    client: UnarySessionClient<Channel>,
+}
+
+impl AttestationClient {
+    pub async fn create(
+        uri: &str,
+        expected_tee_measurement: &[u8],
+        server_verifier: ServerIdentityVerifier,
+    ) -> anyhow::Result<Self> {
+        let session_id = 0;
+        let channel = Channel::from_shared(uri.to_string())
+            .context("Couldn't create gRPC channel")?
+            .connect()
+            .await?;
+        let mut client = UnarySessionClient::new(channel);
+
+        let mut handshaker = ClientHandshaker::new(
+            AttestationBehavior::create_peer_attestation(expected_tee_measurement),
+            server_verifier,
+        );
+        let client_hello = handshaker
+            .create_client_hello()
+            .context("Couldn't create client hello message")?;
+
+        let mut response = client
+            .message(UnaryRequest {
```

Comment:
It may be nice to at some point abstract the underlying client here, in something like a trait that allows sending and receiving messages for a session.

Something like

```rust
trait Transport {
  async fn call(session_id: SessionId, req: &[u8]) -> Result<Vec<u8>>;
}
```

or perhaps even better:

```rust
trait Transport {
  type S: Session;
  async fn create_session() -> S;
}

trait Session {
  async fn call(req: &[u8]) -> Result<Vec<u8>>;
}
```

(or some variation on it)

Then we can implement different remote attestation variants on top of this transport.

Also it may even be possible to define the entire remote attestation protocol in such a way that it is agnostic of whether the underlying protocol is unary or streaming. The "interesting" part would be written only once in a struct similar to the one proposed in this PR, but it would be parametrized with a type implementing Transport, of which we will have a streaming version and a unary version. Not sure if it would actually work though, would like to hear your opinion @jul-sh @ipetr0v .